Check TTL in do_lock

joe-iddon opened the pull request:
#### Reference Issues/PRs
<!--Example: Fixes #1234. See also #3456.-->

#### What does this implement or fix?

The symbol list is protected by a storage lock, so only one writer can compact it at a time. This is implemented using a key of type `LOCK` of which there is almost at most one of in the storage backend. Writers will fail early if they see this key.

The method `lock()` implements a check for the TTL of the storage lock in case it was left locked by a failed write (default TTL is one day). But the `try_lock()` method used in the `list_symbols()` operation does not bother to check the TTL. This PR fixes that and adds another unit test.

The consequence of the previous behaviour was quite bad: libraries could get stuck in a permanent state of an uncompacted symbol list. So for tens of thousands of symbols, or frequent writers (each write inserts a new `__add__` to the symbol list), the `list_symbols()` method would grind to a halt since it would never be able to compact the symbols.

Already tested internally.

#### Any other comments?

#### Checklist

<details>
  <summary>
   Checklist for code changes...
  </summary>
 
 - [x ] Have you updated the relevant docstrings, documentation and copyright notice?
 - [x ] Is this contribution tested against [all ArcticDB's features](../docs/mkdocs/docs/technical/contributing.md)?
 - [x ] Do all exceptions introduced raise appropriate [error messages](https://docs.arcticdb.io/error_messages/)?
 - [x ] Are API changes highlighted in the PR description?
 - [x ] Is the PR labelled as enhancement or bug so it appears in autogenerated release notes?
</details>

<!--
Thanks for contributing a Pull Request to ArcticDB! Please ensure you have taken a look at:
 - ArcticDB's Code of Conduct: https://github.com/man-group/ArcticDB/blob/master/CODE_OF_CONDUCT.md
 - ArcticDB's Contribution Licensing: https://github.com/man-group/ArcticDB/blob/master/docs/mkdocs/docs/technical/contributing.md#contribution-licensing
-->

vasil-pashov commented on cpp/arcticdb/util/test/test_storage_lock.cpp:
```diff
@@ -182,6 +182,46 @@ struct ForceReleaseLockTask {
+        } else {
+            ++data_->vol_;
+            // This should be unnecessary as we are already locked
+            std::lock_guard l{data_->mutex_};
```
`LockData::atomic_` is of type `std::atomic<uint64_t>` (which is provides the strongest memory ordering (Sequentially-consistent ordering)). There is not need to guard touching this variable with a lock. It is guaranteed by the standard that no data race is happening.

joe-iddon replied:
Please see my comment below about scrapping this mutex.

vasil-pashov commented on cpp/arcticdb/util/test/test_storage_lock.cpp:
```diff
@@ -242,3 +285,39 @@ TEST(StorageLock, ForceReleaseLock) {
+    // This is set in nanoseconds => 200ms for the TTL
+    ConfigsMap::instance()->set_int("StorageLock.TTL", 200 * 1000 * 1000);
+
+    std::vector<std::shared_ptr<StorageLock<>>> locks;
```
Where is this used?

joe-iddon replied:
Ah thanks. I noticed it was unused and deleted only one of the two places.

vasil-pashov commented on cpp/arcticdb/util/test/test_storage_lock.cpp:
```diff
@@ -182,6 +182,46 @@ struct ForceReleaseLockTask {
+    }
+
+    folly::Future<folly::Unit> operator()() {
+        std::this_thread::sleep_for(std::chrono::milliseconds(1));
```
Why is there a sleep here?

joe-iddon replied:
Again I'm not sure, just how these unit tests were originally implemented. Agree we should remove then?

vasil-pashov replied:
Seems unnecessary to me. Maybe someone with more context can take a look as well.

joe-iddon replied:
I think fine to remove. The tests still pass and there's no reason to have them.

vasil-pashov commented on cpp/arcticdb/util/test/test_storage_lock.cpp:
```diff
@@ -182,6 +182,46 @@ struct ForceReleaseLockTask {
+            data_->timedout_ = true;
+            data_->contended_ = true;
+        } else {
+            ++data_->vol_;
```
Volatile does not protect from data races, moreover volatile variables are not atomic and must be protected by a lock when concurrent reads and writes are done. The only thing that volatile does is to prevent the compiler from optimizing the value away it has no relation to the (concurrent) memory model. In the test the check data_->vol_ == 4 will produce a flaky test.

joe-iddon replied:
The idea is to have a volatile member that we *want* to be corrupted/not incremented correctly in the case that the storage lock is not doing its job properly.

The atomic one is then a comparison, since this should definitely be updated correctly due to the nature of atomic type in C++.

I agree we don't need to wrap the atomic in a mutex, that is just how it was always done, and I'm not sure why. I think it's a mistake. If you agree should I remove this mutex? I.e. all uses of `data_->mutex_`?

vasil-pashov replied:
> The idea is to have a volatile member that we want to be corrupted/not incremented correctly in the case that the storage lock is not doing its job properly.

I'm not following on this. In the test you're doing the same check for the vol_ as for atmoic_. If the lock is not doing the correct thing vol_ will be a random value between 1 and 4

>  If you agree should I remove this mutex?
Remove it in all places where it guards the atomic, just make sure it's not guarding anything else.

vasil-pashov commented:
@willdealtry I see you're the creator LockData. I have a few questions:
1. timedout_ and contended_ are non-volatile, non-atomic but in several places of this file we're doing concurrent writes. It's most definitely a data race (given the compiler does not omit them due to the lack of volatile). Was this indented? AFAIK on x86 writes are atomic there's no chance of writing half the bool, however I think it's UB to write to them from multiple threads even if they are used just as a flag which can only be set from false to true. Also I'm not quite sure when will the memory be sync-ed, with relaxed atomic the thread join will issue a barrier which will sync the variables. My suggestion is to just use atomic or relaxed atomic (if we care that much about performance)
2. vol_ is an ordinary volatile while atomic_ is atomic. There are a few places where we require vol_ to be the same as atomic_, however vol_ is not guarded by a lock. This does not seem right, AFAIK volatile only prevents the compiler from optimizing the variable away and does not provide any atomicity nor memory ordering. I don't think it's reasonable to compare both (nor to compare vol_ against a specific value as Joe does). Was this on purpose?

joe-iddon commented:
> @willdealtry I see you're the creator LockData. I have a few questions:
> 
> 1. timedout_ and contended_ are non-volatile, non-atomic but in several places of this file we're doing concurrent writes. It's most definitely a data race (given the compiler does not omit them due to the lack of volatile). Was this indented? AFAIK on x86 writes are atomic there's no chance of writing half the bool, however I think it's UB to write to them from multiple threads even if they are used just as a flag which can only be set from false to true. Also I'm not quite sure when will the memory be sync-ed, with relaxed atomic the thread join will issue a barrier which will sync the variables. My suggestion is to just use atomic or relaxed atomic (if we care that much about performance)
> 2. vol_ is an ordinary volatile while atomic_ is atomic. There are a few places where we require vol_ to be the same as atomic_, however vol_ is not guarded by a lock. This does not seem right, AFAIK volatile only prevents the compiler from optimizing the variable away and does not provide any atomicity nor memory ordering. I don't think it's reasonable to compare both (nor to compare vol_ against a specific value as Joe does). Was this on purpose?

After a discussion (Vasil/Joe):

1. We discussed that the `timeout_` and `contended_` members should be `std::atomic` so that multiple threads can fail the test at the same time and not get UB by all trying to report that at once.
2. After explaining that the whole motivation for these tests is to check the storage lock, we deliberately are checking that the volatile member *is* protected by the storage lock being held. If we get the UB that arises from multiple threads writing to it at the same time, then that is great as the test will hopefully fail and then we will know there's a problem with the storage lock implementation. We also check an atomic member is incremented correctly so that we have another way of debugging the test if it goes wrong.

I did discover that Alex's change to fix Windows complaining that the local storage lock mutex is destructed without being unlocked (deliberately), when compiling a debug build was actually invalidating the purpose of the TTL expiration tests since at the end of each test we were calling unlock on the storage lock anyway, not "forgetting to unlock it" as originally intended for the test. I've fixed this now as well. Detail: Added a test method to the storage lock to allow manually just unlocking the local mutex, but leaving the store's `LOCK` key in place, which is what we would like to happen to simulate, e.g. someone CTRL-C on the Python process whilst compacting the symbol list and leaving the `LOCK` key in place.
